File: src/map_renderer.rs
use crate::models::{Country, Geometry, SelectedCountry};
use dioxus::prelude::*;
use std::collections::HashMap;

pub fn render_map(countries: Signal<Vec<Country>>) -> Element {
    let path_cache = use_memo(move || {
        let mut cache = HashMap::new();
        for country in countries.read().iter() {
            let path = match &country.geometry {
                Geometry::Polygon { coordinates } => coordinates_to_path(coordinates),
                Geometry::MultiPolygon { coordinates } => coordinates.iter()
                    .map(|poly| coordinates_to_path(poly))
                    .collect::<Vec<String>>()
                    .join(" ")
            };
            cache.insert(country.name.clone(), path);
        }
        cache
    });

    rsx! {
        svg {
            width: "100%",
            height: "100%",
            view_box: "-180 -90 360 180",
            {countries.read().iter().map(|country| {
                render_country(country, path_cache())
            })}
        }
    }
}

fn render_country(country: &Country, path_cache: HashMap<String, String>) -> Element {
    let mut selected_country = use_context::<Signal<SelectedCountry>>();

    let path = path_cache.get(&country.name).cloned().unwrap_or_default();

    let is_selected = selected_country.read().0.as_ref().map(|c| c.name == country.name).unwrap_or(false);

    let country_name = country.name.clone();
    let country_clone = country.clone();

    rsx! {
        path {
            class: if is_selected { "selected" } else { "" },
            d: "{path}",
            onclick: move |_| {
                selected_country.set(SelectedCountry(Some(country_clone.clone())));
            },
            title {
                "{country_name}"
            }
        }
    }
}
fn coordinates_to_path(coordinates: &[Vec<[f64; 2]>]) -> String {
    coordinates.iter().enumerate().map(|(i, poly)| {
        let mut path = String::new();
        for (j, [x, y]) in poly.iter().enumerate() {
            if j == 0 {
                path.push_str(&format!("M {} {} ", x, -y));
            } else {
                path.push_str(&format!("L {} {} ", x, -y));
            }
        }
        path.push('Z');
        path
    }).collect::<Vec<String>>()
    .join(" ")
}

File: src/lib/map_renderer.ts
// src/lib/map_renderer.ts
import type { Country, Geometry } from './models';

export function renderMap(countries: Country[], selectedCountry: Country | null, handleCountrySelection: (countryName: string) => void): string {
  const pathCache = new Map<string, string>();
  countries.forEach(country => {
    const path = country.geometry.type === 'Polygon'
      ? coordinatesToPath(country.geometry.coordinates)
      : country.geometry.coordinates.map(poly => coordinatesToPath(poly)).join(" ");
    pathCache.set(country.name, path);
  });

  return `
    <svg width="100%" height="100%" viewBox="-180 -90 360 180">
      ${countries.map(country => renderCountry(country, pathCache, selectedCountry, handleCountrySelection)).join('')}
    </svg>
  `;
}

function renderCountry(country: Country, pathCache: Map<string, string>, selectedCountry: Country | null, handleCountrySelection: (countryName: string) => void): string {
  const path = pathCache.get(country.name) || '';
  const isSelected = selectedCountry?.name === country.name;
  const fillColor = {
    'Been': '#4CAF50', // Green
    'Want': '#FFC107', // Yellow
    'Lived': '#2196F3', // Blue
    'None': '#E0E0E0'  // Light Gray
  }[country.visitStatus];

  return `
    <path
      class="country ${isSelected ? 'selected' : ''}"
      d="${path}"
      fill="${fillColor}"
      onclick="(${handleCountrySelection.toString()})('${country.name}')"
    >
      <title>${country.name}</title>
    </path>
  `;
}

function coordinatesToPath(coordinates: number[][][]): string {
  return coordinates.map(poly => {
    let path = '';
    poly.forEach(([x, y], j) => {
      path += j === 0 ? `M ${x} ${-y} ` : `L ${x} ${-y} `;
    });
    return path + 'Z';
  }).join(" ");
}